test/signal_test.dart

```diff
@@ -0,0 +1,428 @@
+      expect(script.kill(), false);
+    });
+
+    test("does not call signal handler after script exited", () async {
+      var completer = Completer<void>();
+      var killCalls = 0;
+      var script = Script.capture((_) async => await completer.future,
+          onSignal: (signal) {
+        killCalls++;
+        return true;
+      });
+      await pumpEventQueue();
+
+      expect(script.kill(), true);
+
+      completer.complete();
+      await pumpEventQueue();
+
+      expect(script.kill(), false);
+      expect(killCalls, 1);
+    });
+
+    test("catches script error from signal handler", () async {
+      var completer = Completer<void>();
+      var script = Script.capture((_) async => await completer.future,
+          onSignal: (signal) => throw ScriptException('onSignal', 42));
+      await pumpEventQueue();
+
+      expect(script.kill(), false);
+
+      completer.complete();
+      expect(script.done, throwsScriptException(42));
+      expect(script.kill(), false);
+    });
+
+    test("catches exception from signal handler", () async {
+      var completer = Completer<void>();
+      var script = Script.capture((_) async => await completer.future,
+          onSignal: (signal) => throw Exception('oh no!'));
+      await pumpEventQueue();
+
+      expect(script.kill(), false);
+
+      completer.complete();
+      expect(script.done, throwsScriptException(257));
+      expect(script.kill(), false);
+    });
+  });
+
+  test("BufferedScript.capture can capture Script.kill signals", () async {
+    var completer = Completer<void>();
+    var signalStream = StreamController<ProcessSignal>();
+    var script = BufferedScript.capture((_) async {
+      signalStream.stream.listen(print);
+      await completer.future;
+      await signalStream.close();
+      print('bye!');
+    }, onSignal: (signal) {
+      signalStream.sink.add(signal);
+      return true;
+    });
+
+    var done = false;
+    var output = script.output.then((v) {
+      done = true;
+      return v;
+    });
+    await pumpEventQueue();
+
+    expect(script.kill(), true);
+
+    completer.complete();
+    expect(done, false);
+
+    await script.release();
+    expect(await output, 'SIGTERM\nbye!');
+    expect(script.kill(), false);
+  });
+
+  test("interrupts a Script.fromLineTransformer", () async {
+    var script = Script.fromLineTransformer(StreamTransformer.fromHandlers(
+      handleData: (data, sink) => sink.add(data.toUpperCase()),
+    ));
+
+    var events = <String>[];
+    var done = false;
+    script.stdout.lines.listen(events.add, onDone: () => done = true);
+
+    script.stdin.writeln('before');
+    await pumpEventQueue();
+
+    expect(script.kill(), true);
+    expect(script.done, throwsScriptException(143));
+
+    await pumpEventQueue();
+
+    expect(done, true);
+    expect(events, ['BEFORE']);
+    expect(script.kill(), false);
+  });
+
+  test("interrupts a ByteStream pipe", () async {
```

Comment:
Any of these tests with streams should also verify that the underlying stream is canceled.